### PR TITLE
 [Drools-3444] Support monitoring 

### PR DESCRIPTION
**Pull request by lucamolteni**
@mswiderski please review this, it's the first extension I created and I'm not sure it's correct :)

Thank you very much

### CODE REVIEW

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
jenkins test again

File: kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/KieServerImpl.java [hunk 1]
```diff
@@ -136,6 +136,7 @@ public void init() {
         List<KieServerExtension> extensions = sortKnownExtensions();
 
         for (KieServerExtension extension : extensions) {
+            logger.info("{} processing", extension);
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/KieServerImpl.java]: do we really need it as info?

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java [hunk 2]
```diff
@@ -126,6 +138,24 @@ public Response listContainers(@Context HttpHeaders headers,
         return createCorrectVariant(server.listContainers(containerFilter), headers);
     }
 
+    @GET
+    @Path("prometheus")
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java]: maybe better call it metrics instead of prometheus ... more generic and provides bit more info that product/project name that might change.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java [hunk 3]
```diff
@@ -126,6 +138,24 @@ public Response listContainers(@Context HttpHeaders headers,
...
+    @Produces({MediaType.TEXT_PLAIN})
+    public Response getModels() {
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java]: why this endpoint is in KieServerRestImpl and not inside the Prometheus KieServer extension?

File: kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusKieServerExtension.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public String getExtensionName() {
+        return EXTENSION_NAME;
+    }
+
+    @Override
+    public Integer getStartOrder() {
+        return Integer.MAX_VALUE;
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusKieServerExtension.java]: shouldn't this be one of the first to init as others might need access to that extension?
https://github.com/kiegroup/droolsjbpm-integration/pull/1708/files#diff-dd2d9058a692175656a6a2553d7195ddR166

File: kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetrics.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
+package org.kie.server.services.prometheus;
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetrics.java]: missing license header

File: kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetricsDMNListener.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
+package org.kie.server.services.prometheus;
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetricsDMNListener.java]: same here, missing license header

File: kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetricsDroolsListener.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
+package org.kie.server.services.prometheus;
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusMetricsDroolsListener.java]: and another one

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-common/src/main/java/org/kie/server/integrationtests/shared/basetests/RestOnlyBaseIntegrationTest.java [hunk 8]
```diff
@@ -53,9 +51,6 @@
 
     @Override
     protected KieServicesClient createDefaultClient() throws Exception {
-        if (TestConfig.isLocalServer()) {
-            configuration = KieServicesFactory.newRestConfiguration(TestConfig.getKieServerHttpUrl(), null, null);
-        }
```
Comment (mswiderski) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-common/src/main/java/org/kie/server/integrationtests/shared/basetests/RestOnlyBaseIntegrationTest.java]: why is this deleted ?

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/java/org/kie/server/integrationtests/dmn/DMNMetricIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    .socketTimeout(10, TimeUnit.SECONDS)
+                    .build();
+        }
+        WebTarget webTarget = httpClient.target(uriString);
+        webTarget.register(new BasicAuthentication(TestConfig.getUsername(), TestConfig.getPassword()));
+        return webTarget;
+    }
+
+    @Test
+    public void test_evaluateAllWithMetrics() {
```
Comment (mswiderski) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/java/org/kie/server/integrationtests/dmn/DMNMetricIntegrationTest.java]: could be only mean but I'd prefer to not have _ in method names

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/resources/logback-test.xml [hunk 10]
```diff
@@ -18,7 +18,8 @@
   <logger name="org.kie.server" level="INFO" />
   <logger name="org.apache.http" level="WARN" />
   <logger name="org.apache.maven" level="WARN" />
-<!--   
+  <logger name="org.kie.server.services.PrometheusMetricsDMNListener" level="DEBUG" />
```
Comment (mswiderski) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/resources/logback-test.xml]: still needed as debug?

File: kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration-drools/src/main/java/org/kie/server/springboot/autoconfiguration/drools/PrometheusKieServerAutoConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConditionalOnClass({KieServerImpl.class})
+@AutoConfigureAfter({DroolsKieServerAutoConfiguration.class})
```
Comment (mswiderski) [inline kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration-drools/src/main/java/org/kie/server/springboot/autoconfiguration/drools/PrometheusKieServerAutoConfiguration.java]: so is Prometheus extension dependent on Drools extension?

**Comment by lucamolteni** [comment]
> why there is no process monitoring and only drools and dmn?

No particular reason, it was just easier to me to start with Drools and DMN.

@etirelli We should probably integrate also processing WDYT?

File: kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusKieServerExtension.java (hunk 4 above)
Comment (lucamolteni) [reply kie-server-parent/kie-server-services/kie-server-services-prometheus/src/main/java/org/kie/server/services/prometheus/PrometheusKieServerExtension.java]: You're right but I can't see why the extension is running correctly even by being the last. 

Anyway DMN has 47 and Drools 0, KieContainer has -1. 

Can I push Drools a bit further (let's say 5) and put this to 2?

We should use Basic numbers for these (10,20,30...)

File: kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration-drools/src/main/java/org/kie/server/springboot/autoconfiguration/drools/PrometheusKieServerAutoConfiguration.java (hunk 11 above)
Comment (lucamolteni) [reply kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration-drools/src/main/java/org/kie/server/springboot/autoconfiguration/drools/PrometheusKieServerAutoConfiguration.java]: It should be the other way around? Probably the order here is not important

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java (hunk 3 above)
Comment (lucamolteni) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java]: Done

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java (hunk 2 above)
Comment (lucamolteni) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-common/src/main/java/org/kie/server/remote/rest/common/resource/KieServerRestImpl.java]: Done

File: kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/KieServerImpl.java (hunk 1 above)
Comment (lucamolteni) [reply kie-server-parent/kie-server-services/kie-server-services-common/src/main/java/org/kie/server/services/impl/KieServerImpl.java]: Now it's debug

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by etirelli** [comment]
@mswiderski we need monitoring done across the board, meaning jbpm and optaplanner as well, but @lucamolteni was assigned to do first Drools core and DMN as he is familiar with that code.

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
Tested locally is this good to go?

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] KIE :: Execution Server 7.18.0-SNAPSHOT ............ SUCCESS [  9.345 s]
[INFO] KIE :: Execution Server :: API ..................... SUCCESS [ 37.030 s]
[INFO] KIE :: Execution Server :: Common .................. SUCCESS [  5.674 s]
[INFO] KIE :: Execution Server :: Remote .................. SUCCESS [  0.334 s]
[INFO] KIE :: Execution Server :: Remote :: Client ........ SUCCESS [ 33.047 s]
[INFO] KIE :: Execution Server :: Controller .............. SUCCESS [  0.272 s]
[INFO] KIE :: Execution Server :: Controller :: API ....... SUCCESS [  6.529 s]
[INFO] KIE :: Execution Server :: Services ................ SUCCESS [  0.466 s]
[INFO] KIE :: Execution Server :: Services :: Common ...... SUCCESS [ 29.475 s]
[INFO] KIE :: Execution Server :: Remote :: JMS ........... SUCCESS [  1.459 s]
[INFO] KIE :: Execution Server :: Remote :: REST .......... SUCCESS [  0.345 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: Common  SUCCESS [  8.347 s]
[INFO] KIE :: Execution Server :: Services :: jBPM Extension SUCCESS [ 22.968 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: jBPM .. SUCCESS [ 10.987 s]
[INFO] KIE :: Execution Server :: Services :: Prometheus Extension SUCCESS [  0.956 s]
[INFO] KIE :: Execution Server :: Services :: Drools Extension SUCCESS [  1.950 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: Drools  SUCCESS [  5.448 s]
[INFO] KIE :: Execution Server :: Services :: jBPM UI Extension SUCCESS [  7.074 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: jBPM UI SUCCESS [ 11.911 s]
[INFO] KIE :: Execution Server :: Services :: OptaPlanner Extension SUCCESS [  1.166 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: OptaPlanner SUCCESS [  5.029 s]
[INFO] KIE :: Execution Server :: Services :: Case Management Extension SUCCESS [  1.491 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: Case Management SUCCESS [  6.826 s]
[INFO] KIE :: Execution Server :: Services :: DMN Extension SUCCESS [  1.110 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: DMN ... SUCCESS [  3.055 s]
[INFO] KIE :: Execution Server :: Services :: Swagger Extension SUCCESS [  2.027 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: Swagger SUCCESS [  0.626 s]
[INFO] KIE :: Execution Server :: Remote :: REST :: Prometheus SUCCESS [  0.609 s]
[INFO] KIE :: Execution Server :: Services :: OpenShift ... SUCCESS [ 21.657 s]
[INFO] KIE :: Execution Server :: Router .................. SUCCESS [  0.141 s]
[INFO] KIE :: Execution Server :: Router :: Proxy ......... SUCCESS [ 17.790 s]
[INFO] KIE :: Execution Server :: Router :: Client ........ SUCCESS [ 16.143 s]
[INFO] KIE :: Execution Server :: Controller :: Implementation SUCCESS [  8.737 s]
[INFO] KIE :: Execution Server :: Controller :: REST Implementation SUCCESS [  5.766 s]
[INFO] KIE :: Execution Server :: Controller :: Standalone Implementation SUCCESS [  0.501 s]
[INFO] KIE :: Execution Server :: Controller :: Websocket Common Implementation SUCCESS [  3.213 s]
[INFO] KIE :: Execution Server :: Controller :: Websocket Implementation SUCCESS [  4.415 s]
[INFO] KIE :: Execution Server :: Standalone Controller ... SUCCESS [01:09 min]
[INFO] KIE :: Execution Server :: Controller :: Websocket Client Implementation SUCCESS [  2.249 s]
[INFO] KIE :: Execution Server :: Controller :: Client .... SUCCESS [  6.167 s]
[INFO] KIE :: Execution Server :: Controller :: OpenShift . SUCCESS [ 16.361 s]
[INFO] KIE :: Execution Server :: Wars .................... SUCCESS [  0.152 s]
[INFO] KIE :: Execution Server :: Wars :: Distribution Wars SUCCESS [05:45 min]
[INFO] KIE :: Execution Server :: Wars :: Distribution .... SUCCESS [ 17.240 s]
[INFO] KIE :: Execution Server :: Tests ................... SUCCESS [  0.787 s]
[INFO] KIE :: Execution Server :: Tests :: JAX-WS Test Web Service SUCCESS [  1.392 s]
[INFO] KIE :: Execution Server :: Tests :: Common Integration Tests SUCCESS [  2.176 s]
[INFO] KIE :: Execution Server :: Tests :: Drools Integration Tests SUCCESS [  3.146 s]
[INFO] KIE :: Execution Server :: Tests :: jBPM Integration Tests SUCCESS [  2.872 s]
[INFO] KIE :: Execution Server :: Tests :: OptaPlanner Integration Tests SUCCESS [  0.599 s]
[INFO] KIE :: Execution Server :: Tests :: Drools and jBPM Integration Tests SUCCESS [  5.884 s]
[INFO] KIE :: Execution Server :: Tests :: Controller Integration Tests SUCCESS [  2.558 s]
[INFO] KIE :: Execution Server :: Tests :: Policies Integration Tests SUCCESS [  1.010 s]
[INFO] KIE :: Execution Server :: Tests :: Router Integration Tests SUCCESS [  9.028 s]
[INFO] KIE :: Execution Server :: Tests :: Case Id Generator Integration Tests SUCCESS [  9.639 s]
[INFO] KIE :: Execution Server :: Tests :: DMN Integration Tests SUCCESS [  0.659 s]
[INFO] KIE :: Execution Server :: Tests :: Custom extension SUCCESS [  0.214 s]
[INFO] KIE :: Execution Server :: Tests :: Custom extension :: Service SUCCESS [  0.578 s]
[INFO] KIE :: Execution Server :: Tests :: Custom extension :: REST SUCCESS [  0.552 s]
[INFO] KIE :: Execution Server :: Tests :: Custom extension :: Client SUCCESS [  0.561 s]
[INFO] KIE :: Execution Server :: Tests :: Custom extension :: Integration Tests SUCCESS [  9.325 s]
[INFO] KIE :: Execution Server :: Deploy Maven Plugin ..... SUCCESS [  3.639 s]
[INFO] KIE :: Execution Server :: Controller :: Deploy Maven Plugin 7.18.0-SNAPSHOT SUCCESS [  2.773 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13:40 min
[INFO] Finished at: 2019-02-11T17:01:32+01:00
[INFO] ---------------------------------------------------------------

**Comment by lucamolteni** [comment]
jenkins execute full downstream build

**Comment by lucamolteni** [comment]
jenkins test again

**Comment by sutaakar** [comment]
@lucamolteni The extension is disabled by default, is it intended behaviour?

**Comment by lucamolteni** [comment]
@sutaakar yes indeed
